sql/driver/driver_test.go

```diff
@@ -258,3 +259,100 @@ func TestInsecure(t *testing.T) {
 		t.Fatal(err)
 	}
 }
+
+// concurrentIncrements starts two Goroutines in parallel, both of which
+// read the integer stored at the other's key, increment and update their own.
+// It checks that the outcome is serializable, i.e. exactly one of the
+// two Goroutines (the later write) sees the previous write by the other.
+func concurrentIncrements(db *sql.DB, t *testing.T) {
+	var wgStart, wgEnd sync.WaitGroup
+	wgStart.Add(2)
+	wgEnd.Add(2)
+	for i := 0; i < 2; i++ {
+		go func(i int) {
+			defer wgEnd.Done()
+			shouldStartWait := true
+			// Loop until success.
+			for {
+				txn, err := db.Begin()
+				if err != nil {
+					t.Fatal(err)
+				}
+				// Although the SELECT and the UPDATE below can be combined into a
+				// single statement, we prefer them being separate to ensure that
+				// this transaction is truly a multi-statement transaction, providing
+				// plenty of opportunity to mess up serializability.
+				var value int64
+				err = txn.QueryRow(`SELECT v FROM t.kv WHERE k = $1`, (i+1)%2).Scan(&value)
+				if err != nil {
+					_ = txn.Rollback()
+					continue
+				}
+				// Synchronization point to ensure that there is contention.
+				if shouldStartWait {
+					wgStart.Done()
+					wgStart.Wait()
+					shouldStartWait = false
+				}
+				value++
+				if _, err := txn.Exec(`UPDATE t.kv SET v = $2 WHERE k = $1`, i, value); err != nil {
+					_ = txn.Rollback()
+					continue
+				}
+				if err := txn.Commit(); err != nil {
+					continue
+				}
+				// Success.
+				break
+			}
+		}(i)
+	}
+	// Wait for the goroutines to finish.
+	wgEnd.Wait()
+}
+
+// TestConcurrentIncrements is a simple explicit test for serializability
+// for the concrete situation described in:
+// https://groups.google.com/forum/#!topic/cockroach-db/LdrC5_T0VNw
+// This test is a copy of the test in client/... which runs the same
+// test for the KV layer. This Belt and Suspenders test is mostly
+// documentation and adds another layer of confidence that transactions
+// are serializable and performant even at the SQL layer.
+func TestConcurrentIncrements(t *testing.T) {
+	defer leaktest.AfterTest(t)
+	s, db := setup(t)
+	defer cleanup(s, db)
+
+	if _, err := db.Exec(`CREATE DATABASE t`); err != nil {
+		t.Fatal(err)
+	}
+	if _, err := db.Exec(`CREATE TABLE t.kv (k INT PRIMARY KEY, v INT)`); err != nil {
+		t.Fatal(err)
+	}
+	// Convenience loop: Crank up this number for testing this
+	// more often. It'll increase test duration though.
+	for k := 0; k < 5; k++ {
+		// Start with a clean slate.
+		if _, err := db.Exec(`TRUNCATE TABLE t.kv`); err != nil {
+			t.Fatal(err)
+		}
+		if _, err := db.Exec(`INSERT INTO t.kv (k, v) VALUES (0,0),(1,0)`); err != nil {
+			t.Fatal(err)
+		}
+		concurrentIncrements(db, t)
+		var min, max, sum int64
```

Comment:
well we initialize the database here so I thought it made sense to add the conclusion in here. The concurrent increments happen in the other function. If you want me to move this conclusion code there, I will also move the initialization code there.